### ISSUE TITLE
Add FIPS 140-3 indicator API

### ISSUE THREAD

**Issue by neverpanic** [openssl/openssl#18501]
FIPS 140-3 requires an indication on whether an operation was using approved services or not. The [FIPS 140-3 implementation guidelines (section 2.4.C)][140-3ig] provide two basic approaches to doing this: implicit indicators, and explicit indicators.

Implicit indicators are basically the concept of "if the operation passes, it was approved" (in the Example Scenarios table given in FIPS 140-3 IG, that's either "Return code from an approved security service call" in scenario 1 or "A status code indicating the completion of service" in scenario 2). At Red Hat, we were originally aiming for implicit indicators in our copy of OpenSSL.

However, this proved to be a problem, because we wanted to certify a signature service, and FIPS 140-3 requires that a signature service computes the digest to be signed within the boundaries of the FIPS module (i.e., using the `OSSL_FUNC_SIGNATURE_DIGEST_SIGN*` provider functions, not the `OSSL_FUNC_DIGEST*` functions followed by `OSSL_FUNC_SIGNATURE_SIGN`). Since we were planning to certify `fips.so` only, this means that `EVP_PKEY_sign`/`EVP_PKEY_verify` would have to be blocked. Unfortunately, `EVP_SignFinal` uses `EVP_PKEY_sign` internally, but outside of `fips.so` and thus outside of the FIPS module boundary. This means that using implicit indicators in combination with certifying only `fips.so` would require blocking both `EVP_PKEY_sign` and `EVP_SignFinal`, which are the two APIs currently used by most users of OpenSSL for signatures.

`EVP_DigestSign` would be acceptable, but has only been added in 3.0 and is thus not yet widely used.

A workaround for this catch-22 would be explicit indicators so that `EVP_PKEY_sign` and `EVP_SignFinal` can continue to work for now, but FIPS-aware applications can query the explicit indicator to check whether the operation was approved.

It would be nice if such an API was available, but unfortunately it seems to be quite complicated to get this introduced through the entire stack from the provider up to the EVP APIs, especially without breaking ABI.

I've made a rough attempt at sketching out an implementation of `EVP_MD_fips_indicator()` in https://github.com/neverpanic/openssl/commit/a25ca79e9f8ae832fb2b68929aa00b40de718e4d, which could potentially be expanded into the other EVP_* structs. We're also considering a symbol that can be `dlopen(3)`/`dlsym(3)`'d from `fips.so` in https://gitlab.com/redhat/centos-stream/rpms/openssl/-/merge_requests/71, but that's more a short-term hack than a long-term solution. See also https://bugzilla.mozilla.org/show_bug.cgi?id=1710773#c0 on how NSS implemented FIPS indicators.

[140-3ig]: https://csrc.nist.gov/CSRC/media/Projects/cryptographic-module-validation-program/documents/fips%20140-3/FIPS%20140-3%20IG.pdf

**Comment by neverpanic** [comment]
Cc: @beldmit.

**Comment by neverpanic** [comment]
I should have added that `EVP_PKEY_sign` can of course be claimed as a raw signature API, and that can be tested and approved. Unfortunately it seems that for RSA, NIST does not provide test vectors (and thus no certification) of the corresponding raw verification operation.

**Comment by neverpanic** [comment]
Hm, maybe I checked the wrong file on that; I think I checked util/libcrypto.num, but that has 3_0_0 for all symbols. Regardless of how long they've been available, unfortunately it seems like they're not being used widely – at least not widely enough that making `EVP_PKEY_sign()` return an error when used with the FIPS provider would not break many applications.

**Comment by t8m** [comment]
IMO this does not make any sense to me - these existing applications would not be compliant anyway so what would be the point of the indicators? The apps can just switch to non-FIPS mode or be fixed to use DigestSign* APIs.

**Comment by neverpanic** [comment]
You're right that they won't be compliant – however, the fallout is a bit unexpected, e.g., ssh no longer works. I think breaking those functions leads to bad user experience, while having indicators would allow for a smoother transition off the old APIs.

**Comment by beldmit** [comment]
@t8m this is a new FIPS-140-3 concept. The FIPS-compliant applications should have to deal with it. It's a long shot but future FIPS certification will require such an API.

**Comment by kaduk** [comment]
> > EVP_DigestSign would be acceptable, but has only been added in 3.0 and is thus not yet widely used.
> 
> EVP_DigestSignInit/Update/Final have been available for a long time (since 1.0.0). Wouldn't they be acceptable? Not sure when the one-shot `EVP_DigestSign` was added but it appears to be before 3.0. It is present in the 1.1.1 man pages.

Since it goes through to the dedicated `digest_sign_update` provider op, it seems like it shoould work from the policy perspective.  But that doesn't resolve the "existing software doesn't actually use it" issue.

**Comment by paulidale** [comment]
> Since it goes through to the dedicated `digest_sign_update` provider op, it seems like it shoould work from the policy perspective. But that doesn't resolve the "existing software doesn't actually use it" issue.

Which leads to the obvious question: _how far should we bend over backwards to support recent changes in the FIPS guidelines?_

Expecting an application to _just work_ when the FIPS standards update is rather optimistic.  Us having to add a hack to support such applications also seems a bit out there.

**Comment by simo5** [comment]
@paulidale indicators are not a hack, but an explicit requirement of the new FIPS-140-3 standard. We kind of invented the concept of implicit indicators, but having explicit ones is the more standards complaint one.

Note that indicators do not have to be on the API themselves. Indicators identify services, so it can be a completely separate facility whereby the OpenSSL FIPS provider declare what "services" it offers and provide indication about them.

However it is much simpler to tie an indicator to an operation (as represented by an API), as that is the most flexible way going forward.

**Comment by simo5** [comment]
@t8m @paulidale 
Just to move the discussion along this is an example of what indicators could look like:
https://github.com/simo5/openssl/commit/f38ca9f089a17b85a9ab54b39cfc4cb935cef15b

Key aspects:
- it allows to confine FIPS checks to the FIPS Indicator parameter processing
- does not require to block operations that are not fips approved because just some of the parameters are not of NIST liking, or just because NIST does not have an ACVP test 
- does not affect applications that do not care about FIPS compliance
- allows applications that care to check and decide by themselves if they want to block or allow the operation (the app may know better)

But the most important aspect is that this adds no new APIs, just a parameter that can be define for all functions once and then the already existing EVP_PKEY_CTX_get_params()/EVP_PKEY_CTX_gettable_params() handles it all.

We could do this as a downstream patch, but I rather have this available for all vendors so that applications do not need to be patched differently for each vendor's way of implementing FIPS-140-3 restrictions.

**Comment by jvdsn** [comment]
> EVP_DigestSignInit/Update/Final have been available for a long time (since 1.0.0). Wouldn't they be acceptable? Not sure when the one-shot `EVP_DigestSign` was added but it appears to be before 3.0. It is present in the 1.1.1 man pages.

Yes, but these functions don't seem to be available in `fips.so`: 
```
$ strings fips.so | grep EVP_Digest | sort | uniq    
EVP_Digest
EVP_DigestFinal
EVP_DigestFinal_ex
EVP_DigestFinalXOF
EVP_DigestInit
EVP_DigestInit_ex
EVP_DigestInit_ex2
EVP_DigestSignUpdate
EVP_DigestUpdate
EVP_DigestVerifyUpdate
```
Only the `DigestSignUpdate` and `DigestVerifyUpdate` are included, for some reason?

**Comment by nhorman** [comment]
marking as inactive, to be closed at the end of 3.4 dev barring further input

**Comment by paulidale** [comment]
@slontis is making slow progress towards implementing indicators.
Not sure this should be closed unless things are being tracked elsewhere.

**Comment by nhorman** [comment]
@paulidale you're right, this was mislabeled on my part this should be marked for a future minor version

**Comment by simo5** [comment]
@paulidale is this progress public somewhere?
I will be a user of this work early on and I would like to make sure the solution will be compatible with my needs and provide feedback should it not be.

**Comment by paulidale** [comment]
Best to ask @slontis the status.

**Comment by slontis** [comment]
https://github.com/openssl/openssl/pull/23609

**Comment by slontis** [comment]
https://github.com/openssl/openssl/pull/23609 is the design doc..

https://github.com/openssl/openssl/pull/23890 shows how to apply it in code.. I need to break this PR up into parts so that it can be merged back in.
